src/mx25r6435f_driver.c

```diff
@@ -0,0 +1,1145 @@
+  * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+  * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+  * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+  * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+  * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+  * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+  * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+  * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+  *
+  ******************************************************************************
+  */
+
+#ifdef __cplusplus
+ extern "C" {
+#endif
+
+/* Includes ------------------------------------------------------------------*/
+#include "mx25r6435f_driver.h"
+#include "PeripheralPins.h"
+
+/** @addtogroup BSP
+  * @{
+  */
+
+/** @addtogroup STM32L475E_IOT01
+  * @{
+  */
+
+/** @defgroup STM32L475E_IOT01_QSPI QSPI
+  * @{
+  */
+
+/* Private constants --------------------------------------------------------*/ 
+/** @defgroup STM32L475E_IOT01_QSPI_Private_Constants QSPI Private Constants
+  * @{
+  */
+#define QSPI_QUAD_DISABLE       0x0
+#define QSPI_QUAD_ENABLE        0x1
+
+#define QSPI_HIGH_PERF_DISABLE  0x0
+#define QSPI_HIGH_PERF_ENABLE   0x1
+/**
+  * @}
+  */
+/* Private variables ---------------------------------------------------------*/
+
+/** @defgroup STM32L475E_IOT01_QSPI_Private_Variables QSPI Private Variables
+  * @{
+  */
+QSPI_HandleTypeDef QSPIHandle;
+
+/**
+  * @}
+  */
+
+
+/* Private functions ---------------------------------------------------------*/
+
+/** @defgroup STM32L475E_IOT01_QSPI_Private_Functions QSPI Private Functions
+  * @{
+  */
+static uint8_t QSPI_ResetMemory        (QSPI_HandleTypeDef *hqspi);
+static uint8_t QSPI_WriteEnable        (QSPI_HandleTypeDef *hqspi);
+static uint8_t QSPI_AutoPollingMemReady(QSPI_HandleTypeDef *hqspi, uint32_t Timeout);
+static uint8_t QSPI_QuadMode           (QSPI_HandleTypeDef *hqspi, uint8_t Operation);
+static uint8_t QSPI_HighPerfMode       (QSPI_HandleTypeDef *hqspi, uint8_t Operation);
+
+/**
+  * @}
+  */
+
+/* Exported functions ---------------------------------------------------------*/
+
+/** @addtogroup STM32L475E_IOT01_QSPI_Exported_Functions
+  * @{
+  */
+
+/**
+  * @brief  Initializes the QSPI interface.
+  * @retval QSPI memory status
+  */
+uint8_t BSP_QSPI_Init(void)
+{ 
+  QSPIHandle.Instance = QUADSPI;
+
+  /* Call the DeInit function to reset the driver */
+  if (HAL_QSPI_DeInit(&QSPIHandle) != HAL_OK)
+  {
+    return QSPI_ERROR;
+  }
+        
+  /* System level initialization */
+  BSP_QSPI_MspInit();
+  
+  /* QSPI initialization */
+  QSPIHandle.Init.ClockPrescaler     = 2; /* QSPI clock = 80MHz / (ClockPrescaler+1) = 26.67MHz */

```

Comment:
what happens if SystemCoreClock is not 80MHz?
the ClockPrescaler will not be ok?

Reply:
If SystemCoreClock is lower, the QSPI clock will be lower. But if the SystemCoreClock is higher and the prescaler is too low then there is a risk the QSPI clock is higher than the maximum authorized (80MHz). So I modified the configuration to be sure the QSPI clock is always lower than 80MHz.